pkg/controller/util.go

```diff
@@ -326,3 +329,297 @@ func addToMap(m1, m2 map[string]string) map[string]string {
+			Annotations: map[string]string{
+				AnnCreatedBy: "yes",
+			},
+		},
+		Spec: v1.PodSpec{
+			Affinity: &v1.Affinity{
+				NodeAffinity: &v1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+						NodeSelectorTerms: []v1.NodeSelectorTerm{
+							{
+								MatchExpressions: []v1.NodeSelectorRequirement{
+									{
+										Key:      "kubevirt.io/Host-Assisted-Cloning",
+										Operator: v1.NodeSelectorOpIn,
+										Values: []string{"true"},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			Containers: []v1.Container{
+				{
+					Command:		[]string{"/bin/sh"},
+					Name:            CLONER_TARGET_PODNAME,
+					Image:           image,
+					ImagePullPolicy: v1.PullPolicy(pullPolicy),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      ImagePathName,
+							MountPath: CLONER_IMAGE_PATH,
+						},
+						{
+							Name:      SocketPathName,
+							MountPath: CLONER_SOCKET_PATH,
+						},
+					},
+					Args: []string{"-c", "/tmp/script.sh target" },
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+			Volumes: []v1.Volume{
+				{
+					Name: ImagePathName,
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.Name,
+							ReadOnly:  false,
+						},
+					},
+				},
+				{
+					Name: SocketPathName,
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: CLONER_SOCKET_PATH,
+						},
+					},
+
+				},
+			},
+		},
+	}
+	return pod
+}
+
+// checkPVC verifies that the passed-in pvc is one we care about. Specifically, it must have the
+// CloneRequest annotation and it must not already be "in-progress". If the pvc passes these filters
+// then true is returned and the source and the target pods will be created. `AnnCloneRequest` indicates that the
+// pvc is targeted for the cloning job. `AnnCloneInProgress` indicates the  pvc is being processed.
+// Note: there is a race condition where the AnnCloneInProgress annotation is not seen in time and as
+// a result the source and the target pods can be created twice (or more, presumably). To reduce this window
+// a Get api call can be requested in order to get the latest copy of the pvc before verifying
+// its annotations.
+func checkClonePVC(client kubernetes.Interface, pvc *v1.PersistentVolumeClaim, get bool) (ok bool, newPvc *v1.PersistentVolumeClaim, err error) {
+	// check if we have proper AnnCloneRequest annotation on the target pvc
+	if !metav1.HasAnnotation(pvc.ObjectMeta, AnnCloneRequest) {
+		glog.V(Vadmin).Infof("pvc annotation %q not found, skipping pvc\n", AnnCloneRequest)
+		return false, pvc, nil
+	}
+
+	//check if the pvc is being processed
+	if metav1.HasAnnotation(pvc.ObjectMeta, AnnCloneInProgress) {
+		glog.V(Vadmin).Infof("pvc annotation %q exists indicating in-progress or completed, skipping pvc\n", AnnCloneInProgress)
+		return false, pvc, nil
+	}
+
```

Comment:
You should also check that there is no CloneOf annotation as that would indicate that the clone was already taken care of by the provisioner (smart clone).

Reply:
done